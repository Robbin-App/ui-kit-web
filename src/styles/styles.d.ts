import 'styled-components';
import { IBaseRobbinTheme } from './theme';

declare module 'styled-components' {
  export type DefaultTheme = IBaseRobbinTheme;
}
