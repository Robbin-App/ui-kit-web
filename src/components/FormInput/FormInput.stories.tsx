import React, { useState } from 'react';
import { Meta, Story } from '@storybook/react';
import FormInput, { IFormInputProps } from './index';

export default {
  title: 'Components/Form Input',
  component: FormInput,
} as Meta;

const Template: Story<IFormInputProps> = (args) => {
  const [value, setValue] = useState<string>(`${args.value}`);
  return (
    <FormInput
      {...args}
      value={value}
      onChange={(e) => setValue(e.target.value)}
    />
  );
};

export const Default = Template.bind({});
Default.args = {
  value: '',
  label: 'First Name',
  formId: 'first-name-id',
};

export const Filled = Template.bind({});
Filled.args = {
  value: 'Jhon',
  label: 'First Name',
  formId: 'first-name-id-filled',
};

export const Error = Template.bind({});
Error.args = {
  value: 'Jhon',
  label: 'First Name',
  formId: 'first-name-id-filled',
  error: 'Error message',
};

export const Disabled = Template.bind({});
Disabled.args = {
  value: 'Jhon',
  label: 'First Name',
  formId: 'first-name-id-filled',
  disabled: true,
};
