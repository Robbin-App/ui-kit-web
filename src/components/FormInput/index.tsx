import React, { useEffect, useRef, useState } from 'react';
import { FormGroup } from '../FormGroup';

export interface IFormInputProps {
  /**
   * Label of the Input element
   */
  label: string;
  /**
   * Id of the input also used in htmlFor in the label
   */
  formId: string;
  /**
   * Error state
   */
  error?: string | boolean;
  /**
   * Disable the input element
   */
  disabled?: boolean;
  /**
   * Value of the Input element
   */
  value?: string;
  /**
   * onChange event of the Input element
   */
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
}

/**
 * Basic Form Input Component
 */
export const FormInput: React.FunctionComponent<
  IFormInputProps & React.InputHTMLAttributes<HTMLInputElement>
> = ({
  label,
  formId,
  error,
  value,
  onChange,
  disabled,
  className,
  ...props
}) => {
  const inputElement = useRef<HTMLInputElement | null>(null);
  const [useMiniLabel, setUseMiniLabel] = useState<boolean>(false);
  const [isFocus, setIsFocus] = useState<boolean>(false);

  const handleInputFocus = (event: React.FocusEvent<HTMLInputElement>) => {
    props.onFocus && props.onFocus(event);
    setUseMiniLabel(true);
    setIsFocus(true);
  };
  const handleInputBlur = (event: React.FocusEvent<HTMLInputElement>) => {
    props.onBlur && props.onBlur(event);
    if (String(value).length === 0) setUseMiniLabel(false);
    setIsFocus(false);
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    props.onKeyDown && props.onKeyDown(event);
    if (event.key.toLowerCase() === 'escape') {
      inputElement.current?.blur();
    }
  };

  useEffect(() => {
    if (!isFocus) setUseMiniLabel(String(value).length > 0);
  }, [value]);
  return (
    <FormGroup
      label={label}
      formId={formId}
      miniLabel={useMiniLabel}
      isFocus={isFocus}
      disabled={disabled}
      error={error}
      className={className}
    >
      <input
        {...props}
        ref={inputElement}
        placeholder=""
        id={formId}
        disabled={disabled}
        value={value}
        onChange={onChange}
        onBlur={handleInputBlur}
        onFocus={handleInputFocus}
        onKeyDown={handleKeyDown}
      />
    </FormGroup>
  );
};

export default FormInput;
