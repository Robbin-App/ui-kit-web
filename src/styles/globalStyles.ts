import { createGlobalStyle, css } from 'styled-components';
import { normalize } from 'styled-normalize';

export const globalStyles = css`
  ${normalize}
  * {
    box-sizing: border-box;
    font-feature-settings: "lnum";
    &:focus {
      outline: 1px solid ${(props) => props.theme.colors.secondary100};
    }
  }
  body {
    font-family: 'Nunito', sans-serif;;
    color: ${(props) => props.theme.colors.text.globalText}
    background: ${(props) => props.theme.colors.greySurface};
    font-size: 16px;
  }
  html,
  body {
    overflow-x: hidden;
    position: relative;
  }
  
  html {
    scroll-behavior: smooth;
  }
  
  input {
    -webkit-appearance: none;
  }

  input:-webkit-autofill,
  input:-webkit-autofill:hover,
  input:-webkit-autofill:focus textarea:-webkit-autofill,
  textarea:-webkit-autofill:hover textarea:-webkit-autofill:focus,
  select:-webkit-autofill,
  select:-webkit-autofill:hover,
  select:-webkit-autofill:focus {
    -webkit-box-shadow: 0 0 0px 1000px #ffffff inset !important;
  }
  
  h1, h2, h3, h4, h5,h6 {
    color: ${(props) => props.theme.colors.text.headingText}
  }
  
  a {
    color: inherit;
  }
`;

export const GlobalStyles = createGlobalStyle`${globalStyles}`;
