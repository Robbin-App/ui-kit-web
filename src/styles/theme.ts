export interface IBaseRobbinTheme {
  colors: {
    primary90: string;
    primary100: string;
    primary200: string;
    secondary90: string;
    secondary100: string;
    secondary200: string;
    error: string;
    text: {
      headingText: string;
      globalText: string;
      disabledText: string;
    };
    border: {
      darkBorder: string;
      lightBorder: string;
    };
    white: string;
    greySurface: string;
  };
  shadows: {
    global: string;
    globalHover: string;
    focus: string;
    active: string;
  };
  fonts: {
    nunito: string;
    weight: {
      light: number;
      regular: number;
      semiBold: number;
      bold: number;
      extraBold: number;
    };
  };
  sizes: {
    containerMaxWidth: number;
    borderRadius: string;
  };
}

export const baseRobbinTheme: IBaseRobbinTheme = {
  colors: {
    white: '#ffffff',
    greySurface: '#FAFAFA',
    primary90: '#1A5589',
    primary100: '#15456F',
    primary200: '#133D63',
    secondary90: '#5081B9',
    secondary100: '#4677AF',
    secondary200: '#426FA4',
    error: '#A62121',
    border: {
      lightBorder: '#E7E6E6',
      darkBorder: '#D6D6D6',
    },
    text: {
      headingText: '#2E2E2E',
      globalText: '#545454',
      disabledText: '#A1A1A1',
    },
  },
  shadows: {
    global: '0px 0px 9px 2px rgba(0, 0, 0, 0.08)',
    globalHover: '0px 0px 9px 2px rgba(0, 0, 0, 0.16)',
    focus: '0px 0px 9px 2px rgba(21, 69, 111, 0.5)',
    active: '0px 0px 9px 2px rgba(21, 69, 111, 0.8)',
  },
  fonts: {
    nunito: `'Nunito', sans-serif`,
    weight: {
      light: 300,
      regular: 400,
      semiBold: 600,
      bold: 700,
      extraBold: 800,
    },
  },
  sizes: {
    containerMaxWidth: 1230,
    borderRadius: '8px',
  },
};
