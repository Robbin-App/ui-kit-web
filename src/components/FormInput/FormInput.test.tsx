import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { ThemeProvider } from 'styled-components';
import { baseRobbinTheme } from '../../styles';
import FormInput from './index';

const renderPropComponent = (theme, props = {}) => (
  <ThemeProvider theme={theme}>
    <FormInput label="hello input" formId="test-id" {...props} />
  </ThemeProvider>
);

describe('FormInput', () => {
  test('renders', () => {
    const render = renderer
      .create(renderPropComponent(baseRobbinTheme))
      .toJSON();
    expect(render).toMatchSnapshot();
  });

  describe('label', () => {
    test('sets label', () => {
      const { getByLabelText } = render(renderPropComponent(baseRobbinTheme));
      getByLabelText('hello input');
    });
  });

  describe('input', () => {
    test('sets input id', () => {
      const { getByLabelText } = render(renderPropComponent(baseRobbinTheme));
      const input = getByLabelText('hello input', { selector: 'input' });
      expect(input.getAttribute('id')).toEqual('test-id');
    });
    test('sets input value', () => {
      const { getByLabelText } = render(
        renderPropComponent(baseRobbinTheme, {
          value: 'test',
          onChange: jest.fn(),
        })
      );
      const input = getByLabelText('hello input', { selector: 'input' });
      expect(input.value).toEqual('test');
    });
    test('calls onChange on input change', () => {
      const onChange = jest.fn();
      const { getByLabelText } = render(
        renderPropComponent(baseRobbinTheme, {
          value: 'test',
          onChange,
        })
      );
      const input = getByLabelText('hello input', { selector: 'input' });
      fireEvent.change(input, {
        target: { value: 'updated' },
      });
      expect(onChange).toHaveBeenCalledTimes(1);
    });
    test('calls onFocus on input focus', () => {
      const onFocus = jest.fn();
      const { getByLabelText } = render(
        renderPropComponent(baseRobbinTheme, {
          onFocus,
        })
      );
      const input = getByLabelText('hello input', { selector: 'input' });
      fireEvent.focus(input, {
        target: { value: 'updated' },
      });
      expect(onFocus).toHaveBeenCalledTimes(1);
    });
    test('calls onBlur on input blur', () => {
      const onBlur = jest.fn();
      const { getByLabelText } = render(
        renderPropComponent(baseRobbinTheme, {
          onBlur,
        })
      );
      const input = getByLabelText('hello input', { selector: 'input' });
      fireEvent.blur(input, {
        target: { value: 'updated' },
      });
      expect(onBlur).toHaveBeenCalledTimes(1);
    });
    test('calls onKeyDown on input keydown', () => {
      const onKeyDown = jest.fn();
      const { getByLabelText } = render(
        renderPropComponent(baseRobbinTheme, {
          onKeyDown,
        })
      );
      const input = getByLabelText('hello input', { selector: 'input' });
      fireEvent.keyDown(input, {
        target: { value: 'updated' },
      });
      expect(onKeyDown).toHaveBeenCalledTimes(1);
    });
  });

  describe('error', () => {
    test('displays error', () => {
      const { getByText } = render(
        renderPropComponent(baseRobbinTheme, { error: 'error-message' })
      );
      getByText('error-message');
    });
  });
});
