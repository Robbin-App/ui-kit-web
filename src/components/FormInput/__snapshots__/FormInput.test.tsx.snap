// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput renders 1`] = `
<div
  className="sc-bdnxRM bIYsmH"
>
  <div
    className="sc-gtsrHT txfIB form-field"
  >
    <label
      htmlFor="test-id"
    >
      hello input
    </label>
    <input
      id="test-id"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder=""
    />
  </div>
</div>
`;
